tiny_dnn/util/product.h

```diff
@@ -516,6 +540,36 @@ void fill(T *dst, size_t size, T value) {
 
 }  // namespace detail
 
+#ifdef CNN_USE_AVX
+// vertically accumulate eight AVX registers into single register.
+template <typename T>
+MUSTINLINE T accumulate(const T &a,
+                        const T &b,
+                        const T &c,
+                        const T &d,
+                        const T &e,
+                        const T &f,
+                        const T &g,
+                        const T &h) {
+  const T ab   = VECTORIZE_TYPE::add(a, b);
+  const T cd   = VECTORIZE_TYPE::add(c, d);
+  const T ef   = VECTORIZE_TYPE::add(e, f);
+  const T gh   = VECTORIZE_TYPE::add(g, h);
+  const T abcd = VECTORIZE_TYPE::add(ab, cd);
+  const T efgh = VECTORIZE_TYPE::add(ef, gh);
+  return VECTORIZE_TYPE::add(abcd, efgh);
+}
+
+// vertically accumulate four AVX registers into single register.
+template <typename T>
+MUSTINLINE T accumulate(const T &a, const T &b, const T &c, const T &d) {
+  const T ab = VECTORIZE_TYPE::add(a, b);
+  const T cd = VECTORIZE_TYPE::add(c, d);
+  return VECTORIZE_TYPE::add(ab, cd);
+}
+
```

Comment:
@beru for a temporary solution I have kept masking outside of accumulate. As in I am accumulating only full blocks of 8 floats / 4 doubles. That can be refactored once product.h is revamped.